### PR TITLE
Window Airbags (also refactors inflatables)

### PR DESCRIPTION
**Pull request by Gandalf2k15**
## About The Pull Request

Adds window airbags to all reinforced windows. These airbags deploy on atom destruction.

image

This means that all reinforced windows will deploy an airbag when they're destroyed, hopefully reducing the effect of uncontrolled depressurization. 

The airbag is dropped and then armed, anchored, and will eventually pop!

https://user-images.githubusercontent.com/9026500/174913210-8b987d88-b797-45c4-bb4a-725c6536ae81.mp4

I have also ENTIRELY refactored inflatable walls, my god, those were SHITCODE IMPERITUS.

Thanks to TauCeti for the blue version of the inflatable wall! (https://github.com/TauCetiStation/TauCetiClassic/pull/8987)

## Changelog

:cl:
add: All reinforced windows now come with airbags, supplied by Armadyne!
refactor: Inflatable walls have been entirely refactored.
/:cl:

### CODE REVIEW

**Comment by Mryourface13** [comment]
But won't people be upset that they can no longer fully vent the station?

**Comment by Gandalf2k15** [comment]
> 

we used to have airlocks in every window(for a long time), but for performance, they're now these

**Comment by Higgin** [comment]
> But won't people be upset that they can no longer fully vent the station?

Turf holes and people being dented with firelocks are bigger problems than broken windows in my experience, but this is basically a concession to SR players being unable to cope with atmos and its threats.

Engineering pops are frequently low. EVA gear is probably less prevalent relative to the pop even with mass-producible MODs, firesuits in lockers, etc. just because of how many people are on the station. The level of general competence and initiative about fixing atmos issues is probably pretty low (in part due to unfamiliarity, which this will contribute to in some part) - consider how inflatables are widely-available already and so seldom used when, short of grabbing some metal to make a fix, they're two or three quick clicks to stop or slow a breach.

Even without an atmos system that's as punishing as something like ZAS, the state of play unfortunately might justify it. If somebody really wants to open a hole, they can come deliberately shoot/stab/deflate these anyway.

**Comment by Mryourface13** [comment]
Yeah those are great points. As mainly just a crew, I honestly have never used the inflatables even though I now know about them. Biggest problem with changes like these is crew awareness, myself obviously included. 

Where atmos techs, come help us

File: modular_skyrat/modules/window_airbags/code/window_airbag.dm [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+/datum/element/airbag/proc/on_examine(datum/source, mob/user, list/examine_text)
+	SIGNAL_HANDLER
+
+	examine_text += span_warning("It has a blinking red light indicating an airbag is primed and ready to trigger on harsh impact.")
+
+
+
+// A fun little gadget!
+/obj/item/airbag
```
Comment (Zonespace27) [inline modular_skyrat/modules/window_airbags/code/window_airbag.dm:56]: maybe add the `wires` datum to this to allow it to be hacked and at very least detonated early, if not disarmed?

File: modular_skyrat/modules/window_airbags/code/window_airbag.dm [hunk 2]
```diff
@@ -0,0 +1,101 @@
+/obj/structure/window/reinforced/fulltile/Initialize(mapload, direct)
```
Comment (Zonespace27) [inline modular_skyrat/modules/window_airbags/code/window_airbag.dm:3]: So even if an engi creates a window, it'll make the airbag element?

Comment (Gandalf2k15) [reply modular_skyrat/modules/window_airbags/code/window_airbag.dm:3]: Yep, seems okay?

File: modular_skyrat/modules/window_airbags/code/window_airbag.dm (hunk 1 above)
Comment (Gandalf2k15) [reply modular_skyrat/modules/window_airbags/code/window_airbag.dm:56]: Well, I'll probably make it a wrench to remove tbh

File: modular_skyrat/modules/inflatables/code/inflatable.dm [hunk 3]
```diff
@@ -1,120 +1,81 @@
...
-		var/obj/structure/inflatable/R = new structuretype(user.loc)
-		transfer_fingerprints_to(R)
-		R.add_fingerprint(user)
-		qdel(src)
+#define TAPE_REQUIRED_TO_FIX 2
+
 
 /obj/structure/inflatable
 	name = "inflatable wall"
 	desc = "An inflated membrane. Do not puncture."
```
Comment (Gandalf2k15) [inline modular_skyrat/modules/inflatables/code/inflatable.dm]:
```suggestion
	desc = "An inflated membrane. Do not puncture. Alt+Click to deflate."
```

File: modular_skyrat/modules/inflatables/code/inflatable.dm [hunk 4]
```diff
@@ -1,120 +1,81 @@
...
-	attack_generic(M, M.melee_damage_upper)
+	deflate(FALSE)
 
-/obj/structure/inflatable/attack_slime(mob/user)
-	attack_generic(user, rand(10, 15))
+/obj/structure/inflatable/play_attack_sound(damage_amount, damage_type, damage_flag)
+	playsound(src, hit_sound, 75, TRUE)
+
+/obj/structure/inflatable/can_atmos_pass(turf/target_turf, vertical)
+	. = ..()
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/inflatables/code/inflatable.dm]: What's the purpose of this, exactly?

File: modular_skyrat/modules/inflatables/code/inflatable.dm [hunk 5]
```diff
@@ -1,120 +1,81 @@
...
+		return
+	balloon_alert_to_viewers("rapidly deflates!")
+	if(torn_type)
+		new torn_type(get_turf(src))
+	qdel(src)
+
+/obj/structure/inflatable/proc/slow_deflate_finish()
+	if(deflated_type)
+		new deflated_type(get_turf(src))
 	qdel(src)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/inflatables/code/inflatable.dm]: Some small amount of documentation for these two procs would be nice.

File: modular_skyrat/modules/inflatables/code/inflatable.dm [hunk 6]
```diff
@@ -124,118 +85,127 @@
...
-
-/obj/structure/inflatable/door/attack_hand(mob/user)
-	return TryToSwitchState(user)
-/*
-/obj/structure/inflatable/door/CanPass(atom/movable/mover, turf/target, height=0, air_group=0)
+	base_icon_state = "door"
+	torn_type = /obj/item/inflatable/door/torn
+	deflated_type = /obj/item/inflatable/door
+	/// Are we open(FALSE), or are we closed(TRUE)?
+	var/door_state = TRUE
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/inflatables/code/inflatable.dm]: That should be two new defines, being `INFLATABLE_DOOR_OPENED` (0) and `INFLATABLE_DOOR_CLOSED` (1), that'd be less confusing.

File: modular_skyrat/modules/inflatables/code/inflatable.dm [hunk 7]
```diff
@@ -124,118 +85,127 @@
...
+	. = ..()
+	if(torn)
+		to_chat(user, span_warning("[src] is too damaged to function!"))
+		return
+	if(locate(structure_type) in get_turf(user))
+		to_chat(user, span_warning("There is already a wall here!"))
+		return
+	playsound(loc, 'sound/items/zip.ogg', 75, 1)
+	to_chat(user, span_notice("You inflate [src]."))
+	if(do_mob(user, src, 10))
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/inflatables/code/inflatable.dm]:
```suggestion
	if(do_mob(user, src, 1 SECONDS))
```

File: modular_skyrat/modules/inflatables/code/inflatable.dm [hunk 8]
```diff
@@ -124,118 +85,127 @@
...
+	. = ..()
+	icon_state = "[base_icon_state]_[torn ? "torn" : ""]"
+
+/obj/item/inflatable/examine(mob/user)
+	. = ..()
+	if(torn)
+		. += span_warning("It is badly torn, and cannot be used! The damage looks like it could be repaired with some <b>tape</b>.")
+
+/obj/item/inflatable/suicide_act(mob/living/user)
+	visible_message(user, span_danger("[user] starts shoving the [src] up [user.p_their()] ass! It looks like [user.p_their()] going to pull the cord, oh shit!"))
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/inflatables/code/inflatable.dm:191]: I think it usually says "It looks like they're trying to commit suicide!" at the end of these, no? Not that it matters for us, as we don't have mechanical suicides.

File: modular_skyrat/modules/inflatables/code/inflatable.dm [hunk 9]
```diff
@@ -255,9 +225,4 @@
 	for(var/i = 0, i < 16, i ++)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/inflatables/code/inflatable.dm]: The 8 and the 16 should probably be defines.

File: modular_skyrat/modules/window_airbags/code/window_airbag.dm [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+/datum/element/airbag/Detach(datum/target)
+	UnregisterSignal(target, list(COMSIG_ATOM_DESTRUCTION, COMSIG_PARENT_EXAMINE))
+
+/datum/element/airbag/proc/deploy_airbag(atom/movable/destroying_atom, damage_flag)
+	SIGNAL_HANDLER
+
+	new airbag_type(get_turf(destroying_atom))
+
+/datum/element/airbag/proc/on_examine(datum/source, mob/user, list/examine_text)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/window_airbags/code/window_airbag.dm:36]: Some documentation could be nice here, especially to explain that you're modifying a list without returning it, which is unusual behavior.

File: modular_skyrat/modules/window_airbags/code/window_airbag.dm [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+	SIGNAL_HANDLER
+
+	examine_text += span_warning("It has a blinking red light indicating an airbag is primed and ready to trigger on harsh impact.")
+
+
+
+// A fun little gadget!
+/obj/item/airbag
+	name = "airbag"
+	desc = "A small package with an explosive package attached. Stand clear!"
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/window_airbags/code/window_airbag.dm]: `A small package with an explosive package attached.`

I feel like we could come up with a different word than package for one of those two, albeit none of it come to mind at the moment.

File: modular_skyrat/modules/window_airbags/code/window_airbag.dm [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+/obj/item/airbag/proc/arm()
+	if(armed)
+		return
+	balloon_alert_to_viewers("armed!")
+	if(!anchored)
+		addtimer(CALLBACK(src, .proc/deploy_anchor), 1 SECONDS)
+	addtimer(CALLBACK(src, .proc/bang), detonate_time, TIMER_CLIENT_TIME)
+	armed = TRUE
+	playsound(src, armed_sound, 100)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/window_airbags/code/window_airbag.dm]: I got a feeling that's going to be loud.

File: modular_skyrat/modules/window_airbags/code/window_airbag.dm [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+	. = ..()
+	icon_state = "[base_icon_state]_[armed ? "armed" : "safe"]"
+
+/obj/item/airbag/proc/arm()
+	if(armed)
+		return
+	balloon_alert_to_viewers("armed!")
+	if(!anchored)
+		addtimer(CALLBACK(src, .proc/deploy_anchor), 1 SECONDS)
+	addtimer(CALLBACK(src, .proc/bang), detonate_time, TIMER_CLIENT_TIME)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/window_airbags/code/window_airbag.dm]: Where's `TIMER_CLIENT_TIME` defined?

File: modular_skyrat/modules/window_airbags/code/window_airbag.dm [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+	if(!isturf(loc) || anchored)
+		return
+	balloon_alert_to_viewers("anchor deployed!")
+	anchored = TRUE
+
+/obj/item/airbag/proc/bang()
+	var/obj/created_object = new drop_type(get_turf(src))
+	playsound(src, bang_sound, 100, pressure_affected = FALSE)
+	do_smoke(1, 1, created_object, get_turf(src))
+	qdel(src)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/window_airbags/code/window_airbag.dm]: Documentation here could be cool

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]
My one question here: What happens if someone deconstructs it, rather than destroying it? Does that still cause the airbag to expand?

Other than that, I think this is a much better solution than the window firelocks, good shout.

File: modular_skyrat/modules/window_airbags/code/window_airbag.dm (hunk 10 above)
Comment (Gandalf2k15) [reply modular_skyrat/modules/window_airbags/code/window_airbag.dm:36]: This is how it's done, the examine signal works like this

File: modular_skyrat/modules/window_airbags/code/window_airbag.dm (hunk 13 above)
Comment (Gandalf2k15) [reply modular_skyrat/modules/window_airbags/code/window_airbag.dm]: That shouldn't be there actually.

File: modular_skyrat/modules/inflatables/code/inflatable.dm (hunk 4 above)
Comment (Gandalf2k15) [reply modular_skyrat/modules/inflatables/code/inflatable.dm]: it has none

File: modular_skyrat/modules/inflatables/code/inflatable.dm (hunk 8 above)
Comment (Gandalf2k15) [reply modular_skyrat/modules/inflatables/code/inflatable.dm:191]: yeah this is pretty much unused

**Comment by Iamgoofball** [comment]
:-1:

Atmos techs need to learn the hard way that pumping distro with literally 100% of the air supply roundstart is a bad idea

**Pull request merged**